tests/test_csp_rendering.py

```diff
@@ -0,0 +1,148 @@
+                raise self.failureException(f"{item} has no nonce attribute.")
+
+    def _fail_if_found(self, root: Element, path: str, namespaces: Dict[str, str]):
+        """
+        Search elements, fail if a `nonce` attribute is found on them.
+        """
+        elements = root.findall(path=path, namespaces=namespaces)
+        for item in elements:
+            if "nonce" in item.attrib:
+                raise self.failureException(f"{item} has a nonce attribute.")
+
+    def _fail_on_invalid_html(self, content: bytes, parser: HTMLParser):
+        """Fail if the passed HTML is invalid."""
+        if parser.errors:
+            default_msg = ["Content is invalid HTML:"]
+            lines = content.split(b"\n")
+            for position, error_code, data_vars in parser.errors:
+                default_msg.append("  %s" % E[error_code] % data_vars)
+                default_msg.append("    %r" % lines[position[0] - 1])
+            msg = self._formatMessage(None, "\n".join(default_msg))
+            raise self.failureException(msg)
+
+    @override_settings(
+        MIDDLEWARE=settings.MIDDLEWARE + ["csp.middleware.CSPMiddleware"]
+    )
+    def test_exists(self):
+        """A `nonce` should exist when using the `CSPMiddleware`."""
+        response = self.client.get(path="/regular/basic/")
+        if not isinstance(response, HttpResponse):
+            raise self.failureException(f"{response!r} is not a HttpResponse")
+        self.assertEqual(response.status_code, 200)
+
+        html_root: Element = self.parser.parse(stream=response.content)
+        self._fail_on_invalid_html(content=response.content, parser=self.parser)
+        self.assertContains(response, "djDebug")
+
+        namespaces = get_namespaces(element=html_root)
+        toolbar = list(DebugToolbar._store.values())[0]
+        nonce = str(toolbar.request.csp_nonce)
+        self._fail_if_missing(
+            root=html_root, path=".//link", namespaces=namespaces, nonce=nonce
+        )
+        self._fail_if_missing(
+            root=html_root, path=".//script", namespaces=namespaces, nonce=nonce
+        )
+
+    @override_settings(
+        DEBUG_TOOLBAR_CONFIG={"DISABLE_PANELS": set()},
+        MIDDLEWARE=settings.MIDDLEWARE + ["csp.middleware.CSPMiddleware"],
+    )
+    def test_redirects_exists(self):
+        response = self.client.get("/redirect/")
+        if not isinstance(response, HttpResponse):
+            raise self.failureException(f"{response!r} is not a HttpResponse")
+        self.assertEqual(response.status_code, 200)
+
+        html_root: Element = self.parser.parse(stream=response.content)
+        self._fail_on_invalid_html(content=response.content, parser=self.parser)
+        self.assertContains(response, "djDebug")
+
+        namespaces = get_namespaces(element=html_root)
+        context: ContextList = response.context  # pyright: ignore[reportAttributeAccessIssue]
+        nonce = str(context["toolbar"].request.csp_nonce)
```

Comment:
```suggestion
        toolbar = list(DebugToolbar._store.values())[0]
        nonce = str(toolbar.request.csp_nonce)
```

Reply:
Same issue here as above.

It might be inconvenient for others, but I can assure you I have to fight with these kinda errors throughout everywhere when working with django, even when having the stubs installed.

It is not perfect but it's the best we have. I have already made the compromise to have my code "autocorrected" (in the sense of might makes right), so I suggest others make compromises here too.